cherab/core/model/lineshape.pyx

```diff
@@ -367,6 +425,388 @@ cdef class StarkBroadenedLine(LineShapeModel):
+                             given magnetic field strength.
+    :param str polarisation: Leaves only :math:`\pi`-/:math:`\sigma`-polarised components:
+                             "pi" - leave only :math:`\pi`-polarised components,
+                             "sigma" - leave only :math:`\sigma`-polarised components,
+                             "no" - leave all components (default).
+
+    """
+
+    def __init__(self, Line line, double wavelength, Species target_species, Plasma plasma,
+                 ZeemanStructure zeeman_structure, polarisation='no'):
+
+        super().__init__(line, wavelength, target_species, plasma, polarisation)
+
+        self._zeeman_structure = zeeman_structure
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    @cython.initializedcheck(False)
+    @cython.cdivision(True)
+    cpdef Spectrum add_line(self, double radiance, Point3D point, Vector3D direction, Spectrum spectrum):
+
+        cdef int i
+        cdef double ts, sigma, shifted_wavelength, component_radiance
+        cdef Vector3D ion_velocity
+        cdef double[:, :] multiplet_pi_mv, multiplet_sigma_mv
+
+        ts = self.target_species.distribution.effective_temperature(point.x, point.y, point.z)
+        if ts <= 0.0:
+            return spectrum
+
+        ion_velocity = self.target_species.distribution.bulk_velocity(point.x, point.y, point.z)
+
+        # calculate the line width
+        sigma = thermal_broadening(self.wavelength, ts, self.line.element.atomic_weight)
+
+        # obtain magnetic field
+        b_field = self.plasma.get_b_field().evaluate(point.x, point.y, point.z)
+        b_magn = b_field.get_length()
+
+        if b_magn == 0:
+            # no splitting if magnetic filed strength is zero
+            shifted_wavelength = doppler_shift(self.wavelength, direction, ion_velocity)
+            if self._polarisation == NO_POLARISATION:
+                return add_gaussian_line(radiance, shifted_wavelength, sigma, spectrum)
+
+            return add_gaussian_line(0.5 * radiance, shifted_wavelength, sigma, spectrum)
+
+        # coefficients for intensities parallel and perpendicular to magnetic field
+        cos_sqr = (b_field.dot(direction.normalise()) / b_magn)**2
+        sin_sqr = 1. - cos_sqr
+
+        # adding pi components of the Zeeman multiplet
+        if self._polarisation != SIGMA_POLARISATION:
+            component_radiance = 0.5 * sin_sqr * radiance
+            multiplet_mv = self._zeeman_structure.evaluate(b_magn, PI_POLARISATION)
+
+            for i in range(multiplet_mv.shape[1]):
+                shifted_wavelength = doppler_shift(multiplet_mv[MULTIPLET_WAVELENGTH, i], direction, ion_velocity)
+                spectrum = add_gaussian_line(component_radiance * multiplet_mv[MULTIPLET_RATIO, i], shifted_wavelength, sigma, spectrum)
+
+        # adding sigma components of the Zeeman multiplet
+        if self._polarisation != PI_POLARISATION:
```

Comment:
Invert logic.

Reply:
Shouldn't this be `else if` for safety?

Reply:
Unlike `ZeemanStructure.evaluate()` there are three options here. We could have pi-polarised components, sigma-polarised components or both if we collect our signal without a polariser. The `ZeemanStructure.evaluate()` in turn must return pi- and sigma-polarised components separately because the ratio between pi- and sigma-polarised components is determined in the `add_line()` by the angle between the magnetic field vector and the ray.